File: AgqrPlayer4Tv/AgqrPlayer4Tv/Activities/MainActivity.cs
using System;
using AgqrPlayer4Tv.Components.Fragments;
using AgqrPlayer4Tv.Model.Platform;
using AgqrPlayer4Tv.Services;
using Android.App;
using Android.Content;
using Android.Content.PM;
using Android.OS;
using Android.Runtime;
using Android.Views;
using HockeyApp;
using Reactive.Bindings;

namespace AgqrPlayer4Tv.Activities
{
    [Activity(Label = "@string/ApplicationName", MainLauncher = true, Icon = "@drawable/Icon", LaunchMode = LaunchMode.SingleTop)]
    [IntentFilter(new[] { Intent.ActionMain }, Categories = new[] { Intent.CategoryLeanbackLauncher })]
    public class MainActivity : Activity
    {
        private PlaybackPlayerFragmentBase _playerFragment;

        protected override void OnCreate(Bundle bundle)
        {
            CrashManager.Register(this);

            BootupActivity.ScheduleRecommendationUpdate(ApplicationContext);
            StartService(new Intent(this, typeof(UpdateRecommendationService)));

            // DayDream(スクリーンセーバー)を抑制する
            this.Window.SetFlags(WindowManagerFlags.KeepScreenOn, WindowManagerFlags.KeepScreenOn);

            SetContentView(Resource.Layout.Main);

            UIDispatcherScheduler.Initialize();

            base.OnCreate(bundle);
        }

        protected override void OnResume()
        {
            // プレイヤー設定が変わってるかもしれないので毎度作り直す
            this.SetupPlayerFragment();

            base.OnResume();
        }


        protected override void OnPause()
        {
            ApplicationMain.ServiceLocator.GetInstance<ApplicationPreference>().IsLastShutdownCorrectly.Value = true; // 正常シャットダウンフラグを立てる

            base.OnPause();
        }

        protected override void OnDestroy()
        {
            base.OnDestroy();
        }

        private void SetupPlayerFragment()
        {
            var appPrefs = ApplicationMain.ServiceLocator.GetInstance<ApplicationPreference>();

            var transaction = this.FragmentManager.BeginTransaction();
            var playerType = appPrefs.PlayerType.Value;

            // RTMPは突然の死を迎えることがある…
            if (!appPrefs.IsLastShutdownCorrectly.Value && appPrefs.StreamingType.Value == StreamingType.Rtmp)
            {
                appPrefs.StreamingType.Value = StreamingType.Hls;
            }

            var newFragment = (playerType == PlayerType.ExoPlayer)
                ? (PlaybackPlayerFragmentBase)new PlaybackExoPlayerFragment()
                    : (playerType == PlayerType.AndroidDefault)
                        ? (PlaybackPlayerFragmentBase)new PlaybackDefaultVideoViewFragment()
                        : (PlaybackPlayerFragmentBase)new PlaybackWebViewPlayerFragment();

            if (this._playerFragment != null)
            {
                transaction.Replace(Resource.Id.MainFrame, newFragment, newFragment.GetType().Name);
            }
            else
            {
                transaction.Add(Resource.Id.MainFrame, newFragment, newFragment.GetType().Name);
            }
            transaction.Commit();

            this._playerFragment = newFragment;
            this._playerFragment.Play();

            ApplicationMain.ServiceLocator.GetInstance<ApplicationPreference>().IsLastShutdownCorrectly.Value = false; // 正常シャットダウンフラグを折っておく
        }

    }
}

File: AgqrPlayer4Tv/app/src/main/java/org/misuzilla/agqrplayer4tv/component/activity/MainActivity.kt
package org.misuzilla.agqrplayer4tv.component.activity

import android.os.Bundle
import androidx.fragment.app.FragmentActivity
import android.util.Log
import android.view.KeyEvent
import android.view.WindowManager
import org.misuzilla.agqrplayer4tv.R
import org.misuzilla.agqrplayer4tv.component.fragment.PlaybackExoPlayerFragment
import org.misuzilla.agqrplayer4tv.component.fragment.PlaybackPlayerFragmentBase
import org.misuzilla.agqrplayer4tv.component.fragment.PlaybackWebViewFragment
import org.misuzilla.agqrplayer4tv.component.broadcastreceiver.BootupBroadcastReceiver
import org.misuzilla.agqrplayer4tv.component.service.UpdateRecommendationService
import org.misuzilla.agqrplayer4tv.model.preference.ApplicationPreference
import org.misuzilla.agqrplayer4tv.model.preference.PlayerType
import android.content.Intent
import androidx.lifecycle.Observer
import org.misuzilla.agqrplayer4tv.AgqrPlayerApplication
import org.misuzilla.agqrplayer4tv.model.NowPlaying
import org.misuzilla.agqrplayer4tv.model.Reservation

class MainActivity : FragmentActivity() {
    private var currentFragment: PlaybackPlayerFragmentBase? = null
    private var lastPlayerType: PlayerType = PlayerType.EXO_PLAYER

    override fun onAttachedToWindow() {
        Log.d("MainActivity", "onAttachedToWindow")
        super.onAttachedToWindow()

        updateWindowFlags()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        Log.d("MainActivity", "onCreate")
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        // スケジュールして、同時にサービスを起動する
        ApplicationPreference.setRecommendationCurrentProgram("")
        BootupBroadcastReceiver.scheduleRecommendationUpdate(applicationContext)
        startService(Intent(this, UpdateRecommendationService::class.java))

        NowPlaying.now.getProgram().observe(this, Observer {
            Log.d(AgqrPlayerApplication.TAG, "Current: ${it.title} (${it.mailAddress}) ${it.start}-${it.end}" )
            Reservation.instance.removeAllExpired()
        })
        Reservation.instance.onChangeAsObservable.subscribe {
            startService(Intent(this, UpdateRecommendationService::class.java))
        }

        updateWindowFlags()

        setupPlayerFlagment()
    }

    override fun onResume() {
        Log.d("MainActivity", "onResume")
        super.onResume()

        updateWindowFlags()

        if (lastPlayerType != ApplicationPreference.getPlayerType().value) {
            setupPlayerFlagment()
        }
    }

    override fun onPause() {
        Log.d("MainActivity", "onPause")
        super.onPause()

        clearWindowFlags()

        // 正常シャットダウンフラグを立てる
        ApplicationPreference.setIsLastShutdownCorrectly(true)
    }

    private fun updateWindowFlags() {
        // 起動したらスクリーンをオンにするように
        window.addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED)
        window.addFlags(WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD)
        window.addFlags(WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON)
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    private fun clearWindowFlags() {
        // 起動したらスクリーンをオンにするようにしたやつを落とす
        // そうじゃないとActivityが生きている間、無限につきっぱなしになる
        window.clearFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED)
        window.clearFlags(WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD)
        window.clearFlags(WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON)
        window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON) // DayDream(スクリーンセーバー)を抑制する
    }

    override fun onDestroy() {
        Log.d("MainActivity", "onDestroy")

        super.onDestroy()
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        when (keyCode) {
            KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE -> {
                currentFragment?.let {
                    if (it.viewModel.isPlaying().value!!) {
                        it.stop()
                    } else {
                        it.play()
                    }
                }
                return true
            }
        }

        return super.onKeyDown(keyCode, event)
    }

    fun setupPlayerFlagment() {

        // RTMPは突然の死を迎えることがある…
//        if (!ApplicationPreference.isLastShutdownCorrectly.get() && ApplicationPreference.streamingType.get() == StreamingType.RTMP) {
//            ApplicationPreference.streamingType.set(StreamingType.HLS)
//        }

        val fragment = when (ApplicationPreference.getPlayerType().value) {
            PlayerType.EXO_PLAYER -> PlaybackExoPlayerFragment()
            PlayerType.ANDROID_DEFAULT -> PlaybackExoPlayerFragment()
            PlayerType.WEB_VIEW -> PlaybackWebViewFragment()
            else -> throw NotImplementedError()
        }

        this.supportFragmentManager
                .beginTransaction()
                .replace(R.id.main_frame, fragment)
                .commit()

        // 正常シャットダウンフラグを折っておく
        ApplicationPreference.setIsLastShutdownCorrectly(false)

        lastPlayerType = ApplicationPreference.getPlayerType().value!!
    }
}
